bluemira/equilibria/run.py

```diff
@@ -0,0 +1,460 @@
+        )
+        psi_eof = psi_sof - self.params.tau_flattop.value * self.params.v_burn.value
+        return psi_sof, psi_eof
+
+    def converge_equilibrium(self, eq, problem):
+        """
+        Converge an equilibrium problem from a 'frozen' plasma optimised state.
+        """
+        # TODO: Converge equilibria
+        pass
+
+    def plot(self):
+        """
+        Plot the pulsed equilibrium problem.
+        """
+        n_snapshots = len(self.snapshots)
+        if n_snapshots == 0:
+            return
+
+        f, ax = plt.subplots(1, n_snapshots)
+        for i, (k, snap) in enumerate(self.snapshots.items()):
+            axi = ax[i]
+            snap.eq.plot(ax=axi)
+            snap.coilset.plot(ax=axi)
+            axi.set_title(k)
+        return f
+
+
+class FixedPulsedCoilsetProblem(PulsedCoilsetProblem):
+    """
+    Procedural design for a pulsed tokamak with a known, fixed PF coilset.
+    """
+
+    def __init__(
+        self,
+        params,
+        coilset: CoilSet,
+        grid: Grid,
+        coil_constraints: List[OptimisationConstraint],
+        profiles: Profile,
+        magnetic_targets: MagneticConstraintSet,
+        breakdown_strategy_cls: Type[BreakdownZoneStrategy],
+        breakdown_problem_cls: Type[PremagnetisationCOP],
+        breakdown_optimiser: Optimiser = Optimiser(
+            "COBYLA", opt_conditions={"max_eval": 5000, "ftol_rel": 1e-10}
+        ),
+        breakdown_settings: dict = {"B_stray_con_tol": 1e-8, "n_B_stray_points": 20},
+        equilibrium_problem_cls: Type[CoilsetOptimisationProblem] = MinimalCurrentCOP,
+        equilibrium_optimiser: Optimiser = Optimiser(
+            "SLSQP", opt_conditions={"max_eval": 1000, "ftol_rel": 1e-6}
+        ),
+        equilibrium_settings: Optional[dict] = None,
+    ):
+        self.params = params
+        self.coilset = coilset
+        self.grid = grid
+        self.profiles = profiles
+        self.eq_targets = magnetic_targets
+
+        self._bd_strat_cls = breakdown_strategy_cls
+        self._bd_prob_cls = breakdown_problem_cls
+        self._bd_settings = breakdown_settings
+        self._bd_opt = breakdown_optimiser
+
+        self._eq_prob_cls = equilibrium_problem_cls
+        self._eq_opt = equilibrium_optimiser
+
+        defaults = {
+            "gamma": 1e-8,
+            "target_rms_max": 0.5,
+            "target_rms_con_tol": 1e-6,
+        }
+        self._eq_settings = {**defaults, **equilibrium_settings}
+        self._coil_cons = coil_constraints
+
+        super().__init__()
+
+    def run_premagnetisation(self):
+        """
+        Run the breakdown optimisation problem
+        """
+        R_0 = self.params.R_0.value
+        strategy = self._bd_strat_cls(
+            R_0, self.params.A.value, self.params.tk_sol_ib.value
+        )
+        coilset = deepcopy(self.coilset)
+
+        relaxed = all([c.flag_sizefix for c in coilset.coils.values()])
+        i = 0
+        while i == 0 or not relaxed:
```

Comment:
I think this needs to have a break or raise an error for a configurable certain size of iterations ie `i> num`